### PR TITLE
[Maven-Target] Exclude sources of excluded artifacts too

### PR DESCRIPTION
**Pull request by HannesWell**
Currently sources of excluded artifacts are still contained in the target.

### CODE REVIEW

File: org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenTargetLocation.java [hunk 1]
```diff
@@ -208,9 +208,7 @@ private void generateFeature(TargetBundles bundles, boolean source) throws CoreE
 		} else {
 			bundleFilter = Predicate.not(bundleFilter);
 		}
-		Iterable<TargetBundle> featurePlugins = bundles.bundles() //
-				.filter(e -> !isExcluded(e.getKey()) && !isIgnored(e.getKey()))//
```
Comment (HannesWell) [inline org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenTargetLocation.java:212]: ignored artifacts are not added to TargetBundles.bundles so we don't have to check if they are ignored.

**Comment by laeubi** [comment]
This change seem to do more than than excluding source bundles, we must always add all items to the (internal) state and only exclude them from the target state (therefore the filtering on getBundles / getFeature) because otherwhise the user can't see the items in the UI to (un) exclude them.

File: org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenTargetLocation.java [hunk 2]
```diff
@@ -334,6 +332,9 @@ private File unpackFeatureFile(Artifact artifact, CacheManager cacheManager) {
 
 	private void addBundleForArtifact(Artifact artifact, CacheManager cacheManager, IMaven maven,
 			TargetBundles targetBundles, IProgressMonitor monitor) {
+		if (isExcluded(artifact)) {
```
Comment (laeubi) [inline org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenTargetLocation.java:335]: Excluded items must be part of the `TargetBundles` as we want to still show them as ignored in the UI!

**Comment by HannesWell** [comment]
> This change seem to do more than than excluding source bundles, we must always add all items to the (internal) state and only exclude them from the target state (therefore the filtering on getBundles / getFeature) because otherwhise the user can't see the items in the UI to (un) exclude them.

For the UI the DependencyNode tree seems to be used and even with this change the Editor works as expected, at least for
```
<location includeDependencyDepth="infinite" includeDependencyScopes="compile,runtime" includeSource="true" missingManifest="ignore" type="Maven">
	<dependencies>
		<dependency>
			<groupId>com.google.inject</groupId>
			<artifactId>guice</artifactId>
			<version>7.0.0</version>
			<type>jar</type>
		</dependency>
	</dependencies>
	<exclude>com.google.code.findbugs:jsr305:3.0.1</exclude>
</location>
```

grafik

**Comment by laeubi** [comment]
As said the target bundle is meant to hold the whole state, if one wants to filter out things it should be filtered. e.g. there is `getMavenTargetBundle(Artifact)` that should return the target bundle regardless if it is ignored or not, also `getMavenTargetBundle(MavenTargetDependency dependency)` depends on this and so on.

If you want to filter out the sources of ignored items, then `isIgnored(...)` should be enhanced instead or even better `TargetBundles.bundles.ignoredArtifacts` should also contain the source artifact.

**Pull request merged**